python/cuspatial/cuspatial/io/geoseries_reader.py

```diff
@@ -0,0 +1,284 @@
+                    offsets["polygons"]["rings"].append(
+                        len(poly.exterior.coords) * 2 + rings_current
+                    )
+                    for interior in poly.interiors:
+                        rings_current = offsets["polygons"]["rings"][-1]
+                        offsets["polygons"]["rings"].append(
+                            len(interior.coords) * 2 + rings_current
+                        )
+                        num_rings = num_rings + 1
+                    offsets["polygons"]["polygons"].append(num_rings + current)
+                offsets["polygons"]["mpolys"].append(
+                    len(offsets["polygons"]["polygons"]) - 1
+                )
+        return offsets
+
+    def _read_geometries(self, geoseries, offsets):
+        """
+        Creates a set of buffers sized to fit all of the geometries and
+        iteratively populates them with geometry coordinate values.
+
+        Parameters
+        ----------
+        geoseries : A GeoPandas GeoSeries object.
+        offsets : The set of offsets that correspond to the geoseries argument.
+        """
+        buffers = {
+            "points": cudf.Series(cp.zeros(offsets["points"][-1])),
+            "multipoints": cudf.Series(cp.zeros(offsets["multipoints"][-1])),
+            "lines": cudf.Series(cp.zeros(offsets["lines"][-1])),
+            "polygons": {
+                "polygons": cudf.Series(
+                    cp.zeros(len(offsets["polygons"]["polygons"]))
+                ),
+                "rings": cudf.Series(
+                    cp.zeros(len(offsets["polygons"]["rings"]))
+                ),
+                "coords": cudf.Series(
+                    cp.zeros(offsets["polygons"]["rings"][-1])
+                ),
+            },
+        }
+        read_count = {
+            "points": 0,
+            "multipoints": 0,
+            "lines": 0,
+            "polygons": 0,
+        }
+        inputs = []
+        input_types = []
+        input_lengths = []
+        for geometry in geoseries:
+            if isinstance(geometry, Point):
+                # write a point to the points buffer
+                # increase read_count of points pass
+                p = cp.array(geometry)
+                i = read_count["points"] * 2
+                buffers["points"][i] = p[0]
+                buffers["points"][i + 1] = p[1]
+                read_count["points"] = read_count["points"] + 1
+                input_types.append("p")
+                input_lengths.append(1)
+                inputs.append({"type": "p", "length": 1})
+            elif isinstance(geometry, MultiPoint):
+                points = cp.array(geometry).T
+                size = len(points.T) * 2
+                i = read_count["multipoints"]
+                buffers["multipoints"][slice(i, i + size, 2)] = points[0]
+                buffers["multipoints"][slice(i + 1, i + size, 2)] = points[1]
```

Comment:
here as well

Reply:
Unless `.T` is grossly inefficient in this case, using numpy's `__array_interface__` is desirable. Points in this case is an `n x 2` matrix and I don't want to loop over `n` to copy them out.

Reply:
I eliminated  `.T`, but am still using `cp.array(geometry)` so I can use `__array_interface__` to copy `n` points to GPU.

Reply:
Using `.T` isn't grossly inefficient since it would just change the strides, but having these tiny copies from host to device in this loop here is grossly inefficient.